File: solutions/elixir/boutique-inventory/1/lib/boutique_inventory.ex
defmodule BoutiqueInventory do
  def sort_by_price(inventory) do
    inventory
    |> Enum.sort(&(&1.price <= &2.price))
  end

  def with_missing_price(inventory) do
    inventory
    |> Enum.filter(&(is_nil(&1.price)))
  end

  def increase_quantity(item, count) do
    %{
      name: item.name,
      price: item.price,
      quantity_by_size: item.quantity_by_size
      |> Enum.map(fn {k, v} -> {k, v+count} end)
      |> Enum.into(item.quantity_by_size)
    }
  end

  def total_quantity(item) do
    item.quantity_by_size
    |> Map.values()
    |> Enum.reduce(0, &(&1+&2))
  end
end

File: solutions/ruby/boutique-inventory/1/boutique_inventory.rb
class BoutiqueInventory
  def initialize(items)
    @items = items
  end

  def item_names
    @items.map { |i| i[:name] }.sort
  end

  def cheap
    @items.filter { |i| i[:price] < 30 }
  end

  def out_of_stock
    @items.filter { |i| i[:quantity_by_size].count.zero? }
  end

  def stock_for_item(name)
    @items.find { |i| i[:name] == name }[:quantity_by_size]
  end

  def total_stock
    @items.reduce(0) { |acc, v| v[:quantity_by_size].values.sum + acc }
  end

  private

  attr_reader :items
end
